sway-lib-std/src/u128.sw

```diff
@@ -0,0 +1,310 @@
+    }
+
+    pub fn rsh(self, other: u64) -> Self {
+        // If shifting by at least the number of bits, then saturate with
+        // zeroes.
+        if (other >= 128) {
+            return ~Self::new();
+        };
+
+        // If shifting by at least half the number of bits, then lower word can
+        // be discarded.
+        if (other >= 64) {
+            return ~Self::from(0, self.upper >> (other - 64));
+        };
+
+        // If shifting by less than half the number of bits, then need to
+        // partially shift both upper and lower.
+
+        // Save lowest bits of upper half.
+        let lowest_upper_bits = self.upper << (64 - other);
+
+        let upper = self.upper >> other;
+        let lower = (self.lower >> other) + lowest_upper_bits;
+
+        ~Self::from(upper, lower)
+    }
+}
+
+impl core::ops::Add for U128 {
+    // Add a U128 to a U128. Panics on overflow.
+    pub fn add(self, other: Self) -> Self {
+        let mut upper_128 = self.upper.overflowing_add(other.upper);
+
+        // If the upper overflows, then the number cannot fit in 128 bits, so panic.
+        assert(upper_128.upper == 0);
+        let lower_128 = self.lower.overflowing_add(other.lower);
+
+        // If overflow has occurred in the lower component addition, carry.
+        // Note: carry can be at most 1.
+        if lower_128.upper > 0 {
+            upper_128 = upper_128.lower.overflowing_add(lower_128.upper);
+        };
+
+        // If overflow has occurred in the upper component addition, panic.
+        assert(upper_128.upper == 0);
+
+        U128 {
+            upper: upper_128.lower,
+            lower: lower_128.lower,
+        }
+    }
+}
+
+impl core::ops::Subtract for U128 {
+    // Subtract a U128 from a U128. Panics of overflow.
+    pub fn subtract(self, other: Self) -> Self {
+        // If trying to subtract a larger number, panic.
+        assert(!(self < other));
+
+        let mut upper = self.upper - other.upper;
+        let mut lower = 0;
+
+        // If necessary, borrow and carry for lower subtraction
+        if self.lower < other.lower {
+            lower = ~u64::max() - (other.lower - self.lower - 1);
+            upper = upper - 1;
+        } else {
+            lower = self.lower - other.lower;
+        };
+
+        U128 {
+            upper: upper,
+            lower: lower,
+        }
+    }
+}
+
+impl core::ops::Multiply for U128 {
+    // Multiply a U128 with a U128. Panics of overflow.
+    pub fn multiply(self, other: Self) -> Self {
+        let zero = ~U128::from(0, 0);
+        let one = ~U128::from(0, 1);
+
+        let mut total = ~U128::new();
+        let mut i = 128 - 1 + 1;
+
+        while i > 0 {
+            // Workaround for not having break keyword
+            let shift = i - 1;
+            total = total << 1;
+            if (other & (one << shift)) != zero {
+                total = total + self;
+            }
+
+            i = i - 1;
+        }
```

Comment:
This is obviously pretty expensive and is why it's running out of gas?

Couldn't the algorithm be reduced to `((lhs.upper * rhs.upper) << 64) + (lhs.upper * rhs.lower + lhs.lower * rhs.upper) + lhs.lower * rhs.lower`?  The `*` and `+` operations would need to be able to grab the overflow and use it, so it's a bit trickier than just that, but it would still be cheaper than looping 128 times... maybe?

It would only work if `$of` actually contains the overflowed bits from a multiply, and not just the flag to indicate an overflow... is that how it works?

Reply:
I implemented this and it works for small multiplications that fit in 64 bits, but fails for `~u64::max() * ~u64:;max()`. It looks like a variation of what @simonr0204 had originally implemented for u64*u64 multiplication: https://github.com/FuelLabs/sway/blob/ce5534d6b179203d7efc68e9a550754b3cf24d03/sway-lib-std/src/u128.sw#L120-L146. Can we defer optimizing this to a future PR?

Reply:
Yep, sure.  I may take a look at the IR for a mul and see if there's an easyish optimisation we can employ to get the gas cost down.